controller/state.go

```diff
@@ -11,87 +11,148 @@ import (
 	"github.com/filecoin-project/dealbot/metrics"
 	"github.com/filecoin-project/dealbot/tasks"
 	"github.com/google/uuid"
+	"github.com/libp2p/go-libp2p-core/crypto"
 )
 
 var State *state
 
 type state struct {
-	tasks []*tasks.Task
-	mu    sync.Mutex
+	tasks []*tasks.AuthenticatedTask
+	mu    sync.RWMutex
 }
 
 func (s *state) MarshalJSON() ([]byte, error) {
 	return json.Marshal(s.tasks)
 }
 
-func (s *state) Update(req *client.UpdateTaskRequest, recorder metrics.MetricsRecorder) error {
+func (s *state) Get(UUID string) (*tasks.AuthenticatedTask, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	for _, t := range s.tasks {
+		if t.UUID == UUID {
+			return t, nil
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find task with uuid: %s", UUID)
+}
+
+func (s *state) Update(UUID string, req *client.UpdateTaskRequest, key crypto.PrivKey, recorder metrics.MetricsRecorder) (*tasks.AuthenticatedTask, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	for _, t := range s.tasks {
-		if t.UUID == req.UUID {
+		if t.UUID == UUID {
 			if t.Status == tasks.Available {
 				t.WorkedBy = req.WorkedBy
 				t.StartedAt = time.Now()
 			} else {
 				if t.WorkedBy != req.WorkedBy {
-					return errors.New("task already acquired")
+					return nil, errors.New("task already acquired")
 				}
 			}
 			log.Infow("state update", "uuid", t.UUID, "status", req.Status, "worked_by", req.WorkedBy)
 
 			t.Status = req.Status
+			var err error
+			t.Signature, err = key.Sign(t.Bytes())
```

Comment:
There's no commit message here, so I'm assuming that what you want to do is sign each task with a private key, handing the signature to the daemon grabbing the task, so that only that same daemon can later report updates on the task.

If another client tries to update that task, they would fail as it would be practically impossible for them to guess the same signature.

If my understanding is correct, this feels wrong - we should only create a signature when the task is assigned (next to StartedAt etc), and in the "else" we should verify the signature.

Reply:
nope. this is for external downstream consumers of the dealbot system (bots+controller) to be able to validate that the reports did come from this dealbot system

Reply:
Oh, I see. How do we ensure that only the right client can update a task, then - by assuming that they would otherwise not be able to guess the UUID? I seem to recall that our original UUIDs were really short, though.